### ISSUE TITLE
Recorded Mp4 File getting corrupted.

### ISSUE THREAD

**Issue by SaddamBInSyed** [sskodje/ScreenRecorderLib#83]
Hi @sskodje 
@hakito 
@FoxPride 
@Hondarer 

Thanks for your work.

the application works fine in normal mode.

But whenever I get errors via **OnRecordingFailed** callback then the currently recorded file is getting corrupted.( I can't play in media player or in  chrome).

**Issue 1:**
example scenario:
Application recording the screen for 10 min and next minutes if PC went to screensaver mode then I am getting the error on my error callback **BUT the problem is the whole video is get corrupted.**

**Issue 2:**
Also sometimes **OnRecordingFailed**  **firing with NO ERROR MESSAGE** . 
````
 private static void Rec_OnRecordingFailed(object sender, RecordingFailedEventArgs e)
        {
            try
            {
                log.Debug("Rec_OnRecordingFailed arrived error " + e.Error);     **// e.Error is EMPTY some times.**        
                 isError = true;
                if (!string.IsNullOrWhiteSpace(e.Error))
                {                   
                    if (e.Error.ToLower().Replace(" ", "").Contains("accessisdenied")) //accessisdenied
                    {
                        log.Debug("OnScreenSaverArrived called. stopping and locking");
                        try
                        {
                            rec.Stop();                           
                        }
                        catch { }
                        OnScreenSaverArrived?.Invoke(_current_rec_path, e.Error);                        
                    }
                }
            }
            catch { }
        }`
```


How do I get rid from file corrupt problems even though errors come in between.?

please advise.

**Comment by SaddamBInSyed** [comment]
I noticed that whenever I  **lock/switch user/signout** the PC at that time the current recording file got corrupted.

also noticed that OnRecordingFailed callback fired with NO error message.

Did anyone come across such scenarios? 

please advise.

**Comment by sskodje** [comment]
What is supposed to happen, and does happen when i just tested it, is that when locking, switching desktops, launching fullscreen applications and some other cases, the library will lose access to the desktop for a little while, but it will insert black frames and keep recording until it gains access again. If you get a corrupted video file, it means something crashed. You probably need to provoke a crash and then step through the code and see what happens, or add some logging.

And make sure you are running the latest version.

**Comment by SaddamBInSyed** [comment]
@sskodje 

thank you for the response.

by the way, This (error) behavior is only happening on my windows 10 tablet. But the same scenario is working fine with my Windows 10 PC. ( without any file corruption after lock/switch user etc).

both pc and tablet running the same application with the latest 2.5 version.

any idea or guess why these different results?

Thanks

**Comment by SaddamBInSyed** [comment]
I was tried but since I am new to C++ program I couldn't continue to troubleshoot this issue.

it would be great if you add error log option in your dll code then I can share the error logs at the dll side.

meantime, I would like to ask. Is it possible to save the MP4 file for each frame like as soon as capture the desktop write into the file and save the file.

In this scenario, even DLL get a crash I can able to get the last saved file.

if this code in python /c#, I will definitely add this implementation but C++ may takes months for me.

would like to know your opinion.

**Comment by sskodje** [comment]
Well, if you set "IsFragmentedMp4Enabled" in RecorderOptions to true, it will periodically write mdat data to your video, so it should work even if the recording is aborted. Not every frame, but every few seconds or something like that. But if something crashes, that is undefined behavior, so there is no saying what actually will happen to the file. It would be a lot better to figure out what crashes.

I'll look at adding an option to turn on/off logging and log to a file, but in the meantime you can use [debugview](https://docs.microsoft.com/en-us/sysinternals/downloads/debugview) to check the logs, even if you can't use visual studio.

**Comment by SaddamBInSyed** [comment]
Thanks for the reply.

I will try to set this flag and will do the test again and update you..

**Comment by SaddamBInSyed** [comment]
please find the attached error log which is from debugview console.

[dbgview.log](https://github.com/sskodje/ScreenRecorderLib/files/5116053/dbgview.log)

**Comment by sskodje** [comment]
Can you check if your problem is fixed with 31c2c35302f535862dfaf53aa08e6dcba1dc7f30 ?

**Comment by SaddamBInSyed** [comment]
@sskodje 

Sure. Thank you very much for your time and consideration.

In addition to that, I have tested with enabling "IsFragmentedMp4Enabled" flag and the test result was successful.  recorded mp4 is not corrupting in lock/unlock scenarios. 

Thank you

**Comment by SaddamBInSyed** [comment]
> Can you check if your problem is fixed with [31c2c35](https://github.com/sskodje/ScreenRecorderLib/commit/31c2c35302f535862dfaf53aa08e6dcba1dc7f30) ?

hi @sskodje 
sorry for the late reply.

I have tested this release and its working fine ( video is not corrupted while locking screen).
also, in this release even IsFragmentEnabled = false the video is not getting corrupt.

thank you very much for your prompt response.
